### DISCUSSION TITLE
How to add a absolute positioned Rectangle to VerticalLayout/HorizontalLayout

### DISCUSSION THREAD

**Discussion by wspl** [slint-ui/slint#2684] [Q&A] [answered]
I want to add an overlay or background that can ignore the parent's layout rules to a VerticalLayout/HorizontalLayout. How can I do that?

---

The original scene is that PopupWindow can only be extended by a sized element like VerticalLayout (a Rectange doesn't work). Therefore, I need my `Content` component to inherit VerticalLayout directly to give it a size. But I don't know how to create a background for my `Content` component.

I want to achieve this with the following code:
```qml
component Content inherits VerticalLayout {
  // I want this as a background!
  Rectangle {}
  // ...items...
}

PopupWindow {
    Content {}
}
```

However, I need to change my code to this to make it work.
```qml
PopupWindow {
  Rectangle {}
  VerticalLayout {}
}
```

I'm a web developer. Adding an absolutely positioned element to a container with layout rules is a very common scenario. Perhaps I'm missing some slint-ui features that can achieve this?

---

An additional question:

How can I create a Rectangle that adjusts its size based on its child elements without filling its parent size? I managed to achieve this by wrapping the Rectangle in a HorizontalLayout and using the `alignment` property to prevent it from filling. Is there a more elegant solution for this?

**Comment by ogoffart** [comment, marked as answer]
I think the following should work

```qml

component Content {
  // the background!
  Rectangle {}

  VerticalLayout {
      // ...items...
   }
}
```

> How can I create a Rectangle that adjusts its size based on its child elements without filling its parent size? 

you can use 
```qml
        Rectangle {
            background: green;
            width: self.preferred-width;
            height: self.preferred-height;
            // contents
        }
```

==> [Demo](https://slint-ui.com/releases/1.0.2/editor/?snippet=export+component+Demo+%7B%0A++++width%3A+400px%3B+height%3A+400px%3B%0A++++Rectangle+%7B%0A++++++++background%3A+red%3B%0A++++++++Rectangle+%7B%0A++++++++++++background%3A+green%3B%0A++++++++++++width%3A+self.preferred-width%3B%0A++++++++++++height%3A+self.preferred-height%3B%0A++++++++++++Text+%7B+text%3A+%22hello+world%22%3B+%7D++%0A++++++++%7D%0A++++%7D%0A%7D%0A)

**Reply by wspl** [reply]
I tried the following approach, but it seems to behave differently than when written directly in the PopupWindow. It does not take into account the size and position property in the PopupWindow, but instead centers in a strange place.
```qml
component Content {
  // the background!
  Rectangle {}

  VerticalLayout {
      // ...items...
   }
}
```

The code of PopupWindow:
```qml
popup := PopupWindow {
    x: 0;
    y: root.height;
    width: root.width;

    Content {}
}
```
![image](https://user-images.githubusercontent.com/2133311/236132775-4f68817b-8059-41d2-a52a-4fbc7c34e326.png)
↓
![image](https://user-images.githubusercontent.com/2133311/236132727-5fe079f9-2eb5-4106-9f88-17ca641d8817.png)

**Reply by ogoffart** [reply]
Maybe you need to wrap it in a layout, or write explicitly the height property.

**Reply by wspl** [reply]
Yes, having Content inherit Rectangle as a container, and setting a self.preferred-* size within it, can solve this issue. Thank you for your help.
